scipy/spatial/tests/test_hausdorff.py

```diff
@@ -160,13 +160,17 @@ def test_invalid_dimensions(self):
          # not guaranteed)
          (0.0, 0, 2)),
     ])
-    def test_subsets(self, A, B, seed, expected):
+    def test_subsets(self, A, B, seed, expected, num_parallel_threads):
         # verify fix for gh-11332
         actual = directed_hausdorff(u=A, v=B, seed=seed)
         # check distance
         assert_allclose(actual[0], expected[0])
+        starting_seed = seed
+        if hasattr(seed, 'bit_generator'):
+            starting_seed = seed.bit_generator._seed_seq.entropy
         # check indices
-        assert actual[1:] == expected[1:]
+        if num_parallel_threads == 1 or starting_seed != 77098:
```

Comment:
This is a bit cryptic. The failure for parallel testing with this particular seed looks like:
```
__________________________________________________ TestHausdorff.test_subsets[A5-B5-seed5-expected5] ___________________________________________________
scipy/spatial/tests/test_hausdorff.py:173: in test_subsets
    assert actual[1:] == expected[1:]
E   AssertionError: assert (1, 1) == (0, 2)
E     
E     At index 0 diff: 1 != 0
E     
E     Full diff:
E       (
E     -     0,
E     ?     ^...
E     
E     ...Full output truncated (7 lines hidden), use '-vv' to show
        A          = [(-5, 3), (0, 0)]
        B          = [(0, 1), (0, 0), (-5, 3)]
        actual     = (0.0, 1, 1)
        expected   = (0.0, 0, 2)
        num_parallel_threads = 2
        seed       = Generator(PCG64) at 0x27C647B9840
        self       = <scipy.spatial.tests.test_hausdorff.TestHausdorff object at 0x27c62574390>
        starting_seed = 77098
```

I can't immediately tell why - is there a thread safety issue, or is there a deterministic reason for the mismatch?

Reply:
According to the comment above: https://github.com/scipy/scipy/blob/fd24a7a847e48ac5af2ba92ebdce00b8a27e798e/scipy/spatial/tests/test_hausdorff.py#L158

it states that the indices might change, which I imagine it is what is occurring here

Reply:
Ah makes sense, thanks. @tylerjereddy you're the expert here, so you may want to have a peek at this tweak to the `hausdorff` tests perhaps.

Reply:
This all makes sense as I note below. One question I have is how this relates to "genuine" concurrent implementations of the algorithm, like the Rust one I tried a few years ago in gh-14719 (which also notes the lack of determinism for degenerate inputs)--specifically, this allows threads to run concurrently so that in a given C program like this one ordering guarantees are loosened, but we don't necessarily get substantial performance improvements from that alone right? I.e., we still likely need to write something with atomics/locks in the compiled backend to fully leverage the concurrency/distribute the work and so on?

Is that a reasonable understanding?

Reply:
> specifically, this allows threads to run concurrently so that in a given C program like this one ordering guarantees are loosened

The free-threading work allows Python level threads with the `threading` module to execute in parallel. If those call into C/C++/Cython code, that code will also run in parallel. That parallelism with the default (with-gil) CPython is (a) not possible for Python code, and (b) only possible in C/C++/Cython code if that code explicitly releases the GIL itself, and then re-acquires it before returning the result.

> but we don't necessarily get substantial performance improvements from that alone right?

Correct. For a single function call it doesn't help. The speedups come from the end user starting to use `threading`.

> I.e., we still likely need to write something with atomics/locks in the compiled backend to fully leverage the concurrency/distribute the work and so on?

That's a separate thing entirely. For making a single `hausdorff` call run in parallel, we'd need to apply the `workers=` pattern like done in `scipy.fft` for example. And that may need locks and/or atomics if there are shared data structures, yes.